### PR TITLE
cmd/scollector: WIP for DSC status

### PR DESCRIPTION
**Pull request by gbrayut**
This still needs some error handling, but the basics are there. Going to test it on a few systems tomorrow but would love a code review whenever someone gets a chance.

[Review on Reviewable](https://reviewable.io/reviews/bosun-monitor/bosun/1084)

### CODE REVIEW

**Comment by mjibson** [comment]
Reviewed 9 of 10 files at r1.
Review status: 9 of 10 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 10 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8C3zVwy_ypGB6d-Yu)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L10)):</sup>
No need to name this.

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 86 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8D7wgy3P7hWVE9aCQ)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L86)):</sup>
need to check for error

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 104 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8DaGLvKbnGnxISl4s)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L104)):</sup>
no need for `, _`

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 105 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8Dfxk0TSglUkVe638)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L105)):</sup>
`c.Success++` is fine

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 109 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8DqjIfo2YBvLBAUHv)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L109)):</sup>
same as above

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 123 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8DuHQIzLTdVWjtwFQ)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L123)):</sup>
`.` at EOL

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 130 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8DyA-PfEt5wf3Wil2)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L130)):</sup>
"Time in seconds" or just "Seconds" taken...

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 174 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8E6GV8nL1qGwXqQg6)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L174)):</sup>
```
switch mode {
case "PULL":
  return 0
...
```

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 187 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8EAhi9hF5l_V1I5IB)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L187)):</sup>
```
switch t {
case "Reboot":
  return 0
...
```

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 204 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8ELiNnTKQN5ADNyjp)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L204)):</sup>
nope, gotta return errors all the way up or ignore them. collectors don't ever log errors.

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 207 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8EX9_HEbMNbJzbphd)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L207)):</sup>
I'm not really a fan of using time.Now() for this. Much more inclined to record the actual timestamp and compute things from that instead.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084)

**Comment by mjibson** [comment]
Reviewed 1 of 10 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084)

**Comment by gbrayut** [comment]
Review status: 9 of 10 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 10 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8C3zVwy_ypGB6d-Yu)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L10)):</sup>
Done.

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 86 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8D7wgy3P7hWVE9aCQ)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L86)):</sup>
Done.

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 104 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8DaGLvKbnGnxISl4s)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L104)):</sup>
Done.

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 105 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8Dfxk0TSglUkVe638)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L105)):</sup>
Done.

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 109 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8DqjIfo2YBvLBAUHv)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L109)):</sup>
Done.

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 123 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8DuHQIzLTdVWjtwFQ)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L123)):</sup>
Done.

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 130 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8DyA-PfEt5wf3Wil2)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L130)):</sup>
Done.

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 174 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8E6GV8nL1qGwXqQg6)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L174)):</sup>
Done.

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 204 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8ELiNnTKQN5ADNyjp)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L204)):</sup>
Done.

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 207 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8EX9_HEbMNbJzbphd)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L207)):</sup>
Do you mean run_start_time instead of run_age? Had a discussion about that in slack today https://bosun.slack.com/archives/general/p1434645366001374

I feel pretty strong about this being seconds since and not a timestamp. Lets jump on a hangout tomorrow if you want to talk about it.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084)

**Comment by gbrayut** [comment]
Review status: 9 of 10 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 187 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8EAhi9hF5l_V1I5IB)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L187)):</sup>
Done.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084)

**Comment by mjibson** [comment]
Review status: 9 of 10 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[cmd/scollector/collectors/dsc_windows.go, line 207 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084#-Js8EX9_HEbMNbJzbphd)** ([raw file](https://github.com/bosun-monitor/bosun/blob/2e89f86ffb6310e09fb22338dc9171436b27dfdc/cmd/scollector/collectors/dsc_windows.go#L207)):</sup>
Talk to Kyle about this.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084)

**Comment by mjibson** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bosun-monitor/bosun/1084)

**Comment by gbrayut** [comment]
@mjibson or @captncraig This is ready for a final review before merging. I talked with Kyle about the Age vs Timestamp and he was fine with Age.

File: cmd/scollector/collectors/dsc_windows.go [hunk 1]
```diff
@@ -1,20 +1,29 @@
 package collectors
 
 import (
+	"fmt"
 	"os"
+	"strings"
 	"time"
```
Comment (mjibson) [inline cmd/scollector/collectors/dsc_windows.go]: Fix these imports so they are just in 2 groups: std lib and others.

**Comment by mjibson** [comment]
You can merge this yourself after you fix the imports list.

**Pull request merged**